arc-mlir/src/include/arc/arc-dialect.h

```diff
@@ -23,21 +23,94 @@
 #ifndef ARC_DIALECT_H_
 #define ARC_DIALECT_H_
 
+#include "mlir/IR/StandardTypes.h"
+#include <mlir/IR/Builders.h>
 #include <mlir/IR/Dialect.h>
 #include <mlir/IR/Function.h>
-#include <mlir/IR/Builders.h>
 #include <mlir/IR/Operation.h>
 
 using namespace mlir;
 
 namespace arc {
 
+namespace ArcTypes {
+
+enum Kind {
+  // These kinds will be used by Arc.
+  Appender = Type::Kind::FIRST_PRIVATE_EXPERIMENTAL_0_TYPE,
+};
+
+bool isValueType(Type type);
+
+bool isBuilderType(Type type);
+
+struct AppenderTypeStorage : public TypeStorage {
+  AppenderTypeStorage(Type mergeType) : mergeType(mergeType) {}
+
+  Type mergeType;
+
+  using KeyTy = Type;
+
+  bool operator==(const KeyTy &key) const { return key == KeyTy(mergeType); }
+
+  static llvm::hash_code hashKey(const KeyTy &key) {
+    return llvm::hash_combine(key);
+  }
+
+  static KeyTy getKey(Type mergeType) { return KeyTy(mergeType); }
+
+  static AppenderTypeStorage *construct(TypeStorageAllocator &allocator,
+                                        const KeyTy &key) {
+    return new (allocator.allocate<AppenderTypeStorage>())
+        AppenderTypeStorage(key);
+  }
+};
+
+class AppenderType
+    : public Type::TypeBase<AppenderType, Type, AppenderTypeStorage> {
+public:
+  using Base::Base;
+
+  static bool kindof(ArcTypes::Kind kind) { return kind == ArcTypes::Appender; }
+
+  static AppenderType get(Type mergeType) {
+    mlir::MLIRContext *context = mergeType.getContext();
+    return Base::get(context, ArcTypes::Appender, mergeType);
+  }
+
+  static AppenderType getChecked(Type mergeType, Location location) {
+    mlir::MLIRContext *context = mergeType.getContext();
+    return Base::getChecked(location, context, ArcTypes::Appender, mergeType);
+  }
+
+  static LogicalResult
+  verifyConstructionInvariants(llvm::Optional<Location> loc,
+                               MLIRContext *context, Type type) {
+    //         if (!type.isa<IntegerType>()) {
+    //           if (loc)
+    //             context->emitError(loc) << "non integer-type passed to
+    //             'AppenderType'";
```

Comment:
This code (from the [tutorial](https://mlir.llvm.org/docs/DefiningAttributesAndTypes/)) does not compile. Maybe we could remove the function as it seems to be optional ([Toy](https://mlir.llvm.org/docs/Tutorials/Toy/Ch-7/) does not implement it).

Reply:
Edit: I talked to the MLIR devs, apparently it is deprecated. Instead, we can use `mlir::emitOptionalError`.